benchmarks/htap/lib/controller.py

```diff
@@ -0,0 +1,168 @@
+import os
+import signal
+import time
+import traceback
+
+from contextlib import nullcontext
+from datetime import datetime, timedelta
+from multiprocessing import Manager, Pool, Value
+from pathlib import Path
+from urllib.parse import urlparse
+
+from psycopg2.errors import DuplicateDatabase, DuplicateTable, ProgrammingError
+
+from benchmarks.htap.lib.monitoring import Stats, Monitor
+from benchmarks.htap.lib.queries import Queries
+from benchmarks.htap.lib.transactions import Transactions
+
+from s64da_benchmark_toolkit.dbconn import DBConn
+
+
+def oltp_worker(
+        worker_id, dsn, scale_factor, dry_run, monitoring_interval, reporting_queue, date_value
+):
+    with DBConn(dsn) as conn:
+        start_date = datetime.fromtimestamp(date_value.value)
+        tpcc_worker = Transactions(worker_id, scale_factor, start_date, conn)
+        reporting_interval = timedelta(seconds=monitoring_interval) / 2
+        last_reporting_time = None
+        while True:
+            tpcc_worker.next_transaction(dry_run)
+            now = datetime.now()
+            if last_reporting_time is None or (now - last_reporting_time) >= reporting_interval:
+                reporting_queue.put_nowait(('tpcc', tpcc_worker.stats(worker_id)))
+                last_reporting_time = now
+
+
+def olap_worker(worker_id, dsn, olap_timeout_seconds, dry_run, reporting_queue, date_value):
+    with DBConn(dsn, statement_timeout=olap_timeout_seconds * 1000) as conn:
+        queries = Queries(worker_id, conn)
+        while True:
+            max_date = datetime.fromtimestamp(date_value.value)
+            queries.run_next_query(max_date, reporting_queue, dry_run)
+
+
+class HTAPController:
+    def __init__(self, args):
+        self.dsn = args.dsn
+        self.scale_factor = self._query_num_warehouses()
+        self.dry_run = args.dry_run
+        self.num_oltp_workers = args.oltp_workers
+        self.num_olap_workers = args.olap_workers
+        self.benchmark_duration = args.duration
+        self.olap_query_timeout = args.olap_timeout
+        self.monitoring_interval = args.monitoring_interval
+        self.collect_stats = args.stats_dsn is not None
+        self.stats_dsn = args.stats_dsn
+
+        self.stats = Stats(self.num_oltp_workers, self.num_olap_workers)
+        self.monitor = Monitor(
+                self.stats, self.num_oltp_workers, self.num_olap_workers, self.scale_factor,
+                self.monitoring_interval
+        )
+
+        print(f'Detected scale factor: {self.scale_factor}')
+
+    def _sql_error(self, msg):
+        import sys
+        print(f'ERROR: {msg} Did you run `prepare_benchmark`?')
+        sys.exit(-1)
+
+    def _query_latest_delivery_date(self):
+        with DBConn(self.dsn) as conn:
+            try:
+                conn.cursor.execute('SELECT max(ol_delivery_d) FROM order_line')
+                return conn.cursor.fetchone()[0]
+            except psycopg2.ProgrammingError:
```

Comment:
Does this really work? Because there is no `import psycopg2`

Reply:
I still think that this is an error. Because in order to call `psycopg2.ProgrammingError` we need to import `psycopg2`. Or if we import like `from psycopg2 import ProgrammingError`, then this line should look like `except ProgrammingError`.

Reply:
I am sure I changed this. Are you sure you are looking at the latest version?

Reply:
OK. I thought you didn't.